components/config/gamesettings.cpp

```diff
@@ -173,6 +174,257 @@ bool Config::GameSettings::writeFile(QTextStream &stream)
+    //   v              |                                                |
+    // blank or comment line, save in temp buffer <--------+             |
+    //        |                |                           |             |
+    //        |                +------- comment line ------+             |
+    //        v                    (special processing '##')             |
+    //  "ordered" line                                                   |
+    //        |                                                          |
+    //        v                                                          |
+    // save in a separate map of comments keyed by "ordered" line        |
+    //        |                                                          |
+    //        +----------------------------------------------------------+
+    //
+    //
+    QRegExp settingRegex("^([^=]+)\\s*=\\s*([^,]+)(.*)$");
+    std::vector<QString> comments;
+    std::vector<QString>::iterator commentStart = fileCopy.end();
+    std::map<QString, std::vector<QString> > commentsMap;
+    for (std::vector<QString>::iterator iter = fileCopy.begin(); iter != fileCopy.end(); ++iter)
+    {
+        if (isOrderedLine(*iter))
+        {
+            // save in a separate map of comments keyed by "ordered" line
+            if (!comments.empty())
+            {
+                if (settingRegex.indexIn(*iter) != -1)
+                {
+                    commentsMap[settingRegex.cap(1)+"="+settingRegex.cap(2)] = comments;
+                    comments.clear();
+                    commentStart = fileCopy.end();
+                }
+                // else do nothing, malformed line
+            }
+
+            *iter = QString(); // "ordered" lines to be removed later
+        }
+        else if ((*iter).isEmpty() || (*iter).contains(QRegExp("^\\s*#")))
+        {
+            // comment line, save in temp buffer
+            if (comments.empty())
+                commentStart = iter;
+
+            // special removed content processing
+            if ((*iter).contains(QRegExp("^##content\\s*=")))
+            {
+                if (!comments.empty())
+                {
+                    commentsMap[*iter] = comments;
+                    comments.clear();
+                    commentStart = fileCopy.end();
+                }
+            }
+            else
+                comments.push_back(*iter);
+
+            *iter = QString(); // assume to be deleted later
+        }
+        else
+        {
+            int index = settingRegex.indexIn(*iter);
+
+            // blank or non-"ordered" line, write saved comments
+            if (!comments.empty() && index != -1 && settingRegex.captureCount() >= 2 &&
+                mUserSettings.find(settingRegex.cap(1)) != mUserSettings.end())
+            {
+                for (std::vector<QString>::const_iterator it = comments.begin(); it != comments.end(); ++it)
+                {
+                    *commentStart = *it;
```

Comment:
Coverity-Scan warns about this line.

<pre>
** CID 121666:    (INVALIDATE_ITERATOR)
/components/config/gamesettings.cpp: 296 in Config::GameSettings::writeFileWithComments(QFile &)()
/components/config/gamesettings.cpp: 296 in Config::GameSettings::writeFileWithComments(QFile &)()


________________________________________________________________________________________________________
*** CID 121666:    (INVALIDATE_ITERATOR)
/components/config/gamesettings.cpp: 296 in Config::GameSettings::writeFileWithComments(QFile &)()
290                 // blank or non-"ordered" line, write saved comments
291                 if (!comments.empty() && index != -1 && settingRegex.captureCount() >= 2 &&
292                     mUserSettings.find(settingRegex.cap(1)) != mUserSettings.end())
293                 {
294                     for (std::vector<QString>::const_iterator it = comments.begin(); it != comments.end(); ++it)
295                     {
>>>     CID 121666:    (INVALIDATE_ITERATOR)
>>>     Dereferencing iterator "commentStart" though it is already past the end of its container.
296                         *commentStart = *it;
297                         ++commentStart;
298                     }
299                     comments.clear();
300                     commentStart = fileCopy.end();
301                 }
/components/config/gamesettings.cpp: 296 in Config::GameSettings::writeFileWithComments(QFile &)()
290                 // blank or non-"ordered" line, write saved comments
291                 if (!comments.empty() && index != -1 && settingRegex.captureCount() >= 2 &&
292                     mUserSettings.find(settingRegex.cap(1)) != mUserSettings.end())
293                 {
294                     for (std::vector<QString>::const_iterator it = comments.begin(); it != comments.end(); ++it)
295                     {
>>>     CID 121666:    (INVALIDATE_ITERATOR)
>>>     Dereferencing iterator "commentStart" though it is already past the end of its container.
296                         *commentStart = *it;
297                         ++commentStart;
298                     }
299                     comments.clear();
300                     commentStart = fileCopy.end();
301                 }

</pre>